### PR TITLE
[#435] allow unregister to remove replica by repl_num  or resc_name

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

**Comment by d-w-moore** [comment]
Need test(s).

File: irods/manager/data_object_manager.py [hunk 1]
```diff
@@ -392,9 +398,12 @@ def unlink(self, path, force=False, **options):
 
     def unregister(self, path, **options):
         # https://github.com/irods/irods/blob/4.2.1/lib/api/include/dataObjInpOut.h#L190
-        options[kw.OPR_TYPE_KW] = 26
+        options[kw.OPR_TYPE_KW] = 26 # UNREG_OPR
 
-        self.unlink(path, **options)
+        if {kw.RESC_NAME_KW, kw.REPL_NUM_KW} & set(options.keys()):
```
Comment (trel) [inline irods/manager/data_object_manager.py:403]: let's add a comment here for the 'why' for the next reader

**Comment by d-w-moore** [comment]
Adapting test to run under CI

File: irods/test/data_obj_test.py [hunk 2]
```diff
@@ -1628,6 +1645,42 @@ def test_register_with_xml_special_chars(self):
...
+            # Ensure that two replicas exist.
+            data.replicate(**{kw.DEST_RESC_NAME_KW:uniq2})
+            data = self.sess.data_objects.get(data_path)
+            self.assertEqual(2,len(data.replicas))
+
+            physical_paths = [r.path for r in data.replicas]
+
+            # Assert that unregistering by resource name decrease the number of replicas by 1.
+            data.unregister(**{kw.RESC_NAME_KW:uniq1, kw.COPIES_KW:1})
+            self.assertEqual(1,len(self.sess.data_objects.get(data_path).replicas))
```
Comment (korydraughn) [inline irods/test/data_obj_test.py:1674]: Need to test use of `kw.REPL_NUM_KW` as well.

**Review by alanking — COMMENTED** [review]
Looks good to me. Just had one question for consideration that may or may not be actionable.

File: irods/test/data_obj_test.py [hunk 3]
```diff
@@ -32,15 +33,31 @@
 import irods.parallel
 from irods.manager.data_object_manager import Server_Checksum_Warning
 
-
-def make_ufs_resc_in_tmpdir(session, base_name, allow_local = False):
+def make_ufs_resc_in_tmpdir(session, base_name, allow_local = False, client_vault_mode = (True,)):
+    # Parameters
+    # ----------
+    # base_name         -  The name for the resource, as well as the root directory of the Vault.  Use something random and unlikely to collide.
+    # allow_local       -  Whether to allow the resource's vault to be located under a non-shared ie. "/tmp" style directory.
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods/test/data_obj_test.py:40]: Got it. Carry on.

File: irods/test/data_obj_test.py [hunk 4]
```diff
@@ -32,15 +33,31 @@
 import irods.parallel
 from irods.manager.data_object_manager import Server_Checksum_Warning
 
-
-def make_ufs_resc_in_tmpdir(session, base_name, allow_local = False):
+def make_ufs_resc_in_tmpdir(session, base_name, allow_local = False, client_vault_mode = (True,)):
+    # Parameters
+    # ----------
+    # base_name         -  The name for the resource, as well as the root directory of the Vault.  Use something random and unlikely to collide.
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/data_obj_test.py:62]: Should it be part of the comment / documentation ?

[the remark this answers is not in the document]
Comment (korydraughn) [reply irods/test/data_obj_test.py:62]: I see. The comment made me think you would have to parse the incoming value. The comment really means - _give me a unique string and I'll create a resource and vault using that name_, right?

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/data_obj_test.py:62]: Yes, exactly.  No parsing involved

File: irods/test/data_obj_test.py [hunk 5]
```diff
@@ -32,15 +33,31 @@
...
+    # allow_local       -  Whether to allow the resource's vault to be located under a non-shared ie. "/tmp" style directory.
+    # client_vault_mode -  A tuple of (client_mkdir[, mode_OR_mask]):
+    #                          client_mkdir - whether to call mkdir on the vault-path from the client side, and ...
+    #                          mode_OR_mask - if so, what mode bits to be OR'ed into the permission of the vault path after creation
+    #                                         (A typical value might be : 0o777 | stat.S_ISGID, to guarantee iRODS has permissions on the vault)
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/data_obj_test.py]: I have now put a bit more effort into this to make interface and intent clearer.  Let me know @alanking if you think it's an improvement.

File: irods/test/data_obj_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods/test/data_obj_test.py:62]: Sounds good. Please mark this conversation as resolved.

File: irods/test/data_obj_test.py [hunk 6]
```diff
@@ -32,15 +33,31 @@
...
 from irods.manager.data_object_manager import Server_Checksum_Warning
 
-
-def make_ufs_resc_in_tmpdir(session, base_name, allow_local = False):
+def make_ufs_resc_in_tmpdir(session, base_name, allow_local = False, client_vault_mode = (True,)):
+    # Parameters
+    # ----------
+    # base_name         -  The name for the resource, as well as the root directory of the Vault.  Use something random and unlikely to collide.
+    # allow_local       -  Whether to allow the resource's vault to be located under a non-shared ie. "/tmp" style directory.
+    # client_vault_mode -  A tuple of (client_mkdir[, mode_OR_mask]):
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/data_obj_test.py]: See [conversation ](https://github.com/irods/python-irodsclient/pull/436#discussion_r1144765149) below.

File: irods/test/data_obj_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/data_obj_test.py]: @alanking Do you have an opinion?

[the remark this answers is not in the document]
Comment (alanking) [reply irods/test/data_obj_test.py]: After some discussion, let's just go back to the tuple implementation. We can circle back later if needed.

**Comment by trel** [comment]
tests are passing.

please resolve conversations, squash, and leave off the #

File: irods/test/data_obj_test.py [hunk 7]
```diff
@@ -1665,8 +1665,7 @@ def replica_number_from_resource_name(data_path, resc):
 
             # Use two different ways to specify unregister target:
             for keyword in (kw.RESC_NAME_KW,
-                            kw.REPL_NUM_KW,
-                            ):
+                            kw.REPL_NUM_KW):
```
Comment (trel) [inline irods/test/data_obj_test.py:1668]: or even just fit entire tuple on one line?

Comment (alanking) [reply irods/test/data_obj_test.py:1668]: Agreed

**Comment by d-w-moore** [comment]
> oh, figured it out... - commit message should say...
> 
> "REPL_NAME_KW" -> "REPL_NUM_KW"

ok, took care of that

**Comment by trel** [comment]
squash again please

**Review by korydraughn — APPROVED** [review]

**Comment by d-w-moore** [comment]
Add `#` then?  I count 1 approval and 1 "ship it".  : )

**Comment by trel** [comment]
yes.   please add #.

**Pull request merged**